File: examples/typescript/indexer-authz-v1-query-messages.ts
/**
 * xion.indexer.authz.v1 Query Messages — Protobuf Examples
 *
 * This file demonstrates how to construct, encode, and decode
 * query messages from the Authz Indexer Module (xion.indexer.authz.v1).
 *
 * Queries covered:
 * - QueryGrants: Get grants between specific granter and grantee
 * - QueryGranterGrants: List all grants issued by a granter
 * - QueryGranteeGrants: List all grants received by a grantee
 *
 * The Authz Indexer provides optimized queries for authorization grants,
 * allowing efficient lookup of who can act on behalf of whom.
 *
 * Note: This is an indexer module with query-only messages.
 * To create/revoke grants, use the standard cosmos.authz.v1beta1 module.
 *
 * ⚠️  If you get import errors, run: make proto-gen-ts (in xion-types repo)
 */

import {
  QueryGrantsRequest,
  QueryGrantsResponse,
  QueryGranterGrantsRequest,
  QueryGranterGrantsResponse,
  QueryGranteeGrantsRequest,
  QueryGranteeGrantsResponse,
} from '@burnt-labs/xion-types/types/xion/indexer/authz/v1/query';

import { PageRequest, PageResponse } from '@burnt-labs/xion-types/types/cosmos/base/query/v1beta1/pagination';

// =============================================================================
// QueryGrantsRequest / Response
// =============================================================================
// Get grants between a specific granter and grantee.
// Endpoint: GET /xion/indexer/authz/v1/grants
// Type URL: /xion.indexer.authz.v1.QueryGrantsRequest

function queryGrantsExamples() {
  console.log('=== QueryGrants Examples ===\n');

  // 1. Construct request
  const request = QueryGrantsRequest.fromPartial({
    granter: 'xion1granter...', // Who gave the authorization
    grantee: 'xion1grantee...', // Who received the authorization
    msgTypeUrl: '/cosmos.bank.v1beta1.MsgSend', // Optional: filter by message type
    pagination: {
      key: new Uint8Array(),
      offset: BigInt(0),
      limit: BigInt(10),
      countTotal: true,
      reverse: false,
    },
  });

  console.log('1. Constructed QueryGrantsRequest:');
  console.log('   granter:', request.granter);
  console.log('   grantee:', request.grantee);
  console.log('   msgTypeUrl:', request.msgTypeUrl || '(all types)');
  console.log('   pagination.limit:', request.pagination?.limit?.toString());
  console.log('   typeUrl:', QueryGrantsRequest.typeUrl);

  // 2. Encode request
  const encodedRequest = QueryGrantsRequest.encode(request).finish();
  console.log('\n2. Encoded request:', encodedRequest.length, 'bytes');

  // 3. Decode request
  const decodedRequest = QueryGrantsRequest.decode(encodedRequest);
  console.log('\n3. Decoded granter:', decodedRequest.granter);
  console.log('   Decoded grantee:', decodedRequest.grantee);

  // 4. ProtoMsg
  const protoMsg = QueryGrantsRequest.toProtoMsg(request);
  console.log('\n4. ProtoMsg typeUrl:', protoMsg.typeUrl);

  // === Response ===
  console.log('\n--- Response ---');

  // Note: Grant contains authorization (Any type) and expiration
  const response = QueryGrantsResponse.fromPartial({
    grants: [
      {
        // authorization is an Any type containing the specific authorization
        // e.g., GenericAuthorization, SendAuthorization, StakeAuthorization
        authorization: {
          typeUrl: '/cosmos.authz.v1beta1.GenericAuthorization',
          value: new Uint8Array([10, 30, 47, 99, 111, 115, 109, 111, 115]), // Encoded GenericAuthorization
        },
        expiration: {
          seconds: BigInt(1735689600), // Unix timestamp
          nanos: 0,
        },
      },
    ],
    pagination: {
      nextKey: new Uint8Array(),
      total: BigInt(1),
    },
  });

  console.log('5. Constructed Response:');
  console.log('   grants count:', response.grants.length);
  response.grants.forEach((grant, i) => {
    console.log(`\n   [${i}] Grant:`);
    console.log(`       authorization.typeUrl: ${grant.authorization?.typeUrl}`);
    console.log(`       expiration: ${grant.expiration?.seconds?.toString()} (unix)`);
  });
  console.log('\n   pagination.total:', response.pagination?.total?.toString());
  console.log('   typeUrl:', QueryGrantsResponse.typeUrl);

  // 6. Encode/decode response
  const encodedResponse = QueryGrantsResponse.encode(response).finish();
  const decodedResponse = QueryGrantsResponse.decode(encodedResponse);
  console.log('\n6. Decoded grants count:', decodedResponse.grants.length);

  return { request, response };
}

// =============================================================================
// QueryGranterGrantsRequest / Response
// =============================================================================
// List all grants issued by a specific granter.
// Endpoint: GET /xion/indexer/authz/v1/grants/granter/{granter}
// Type URL: /xion.indexer.authz.v1.QueryGranterGrantsRequest

function queryGranterGrantsExamples() {
  console.log('\n=== QueryGranterGrants Examples ===\n');

  // 1. Construct request
  const request = QueryGranterGrantsRequest.fromPartial({
    granter: 'xion1granter...', // List all grants FROM this address
    pagination: {
      key: new Uint8Array(),
      offset: BigInt(0),
      limit: BigInt(20),
      countTotal: true,
    },
  });

  console.log('1. Constructed QueryGranterGrantsRequest:');
  console.log('   granter:', request.granter);
  console.log('   pagination.limit:', request.pagination?.limit?.toString());
  console.log('   typeUrl:', QueryGranterGrantsRequest.typeUrl);

  // 2. Encode request
  const encodedRequest = QueryGranterGrantsRequest.encode(request).finish();
  console.log('\n2. Encoded request:', encodedRequest.length, 'bytes');

  // 3. ProtoMsg
  const protoMsg = QueryGranterGrantsRequest.toProtoMsg(request);
  console.log('\n3. ProtoMsg typeUrl:', protoMsg.typeUrl);

  // === Response ===
  console.log('\n--- Response ---');

  // GrantAuthorization includes granter, grantee, authorization, and expiration
  const response = QueryGranterGrantsResponse.fromPartial({
    grants: [
      {
        granter: 'xion1granter...',
        grantee: 'xion1grantee1...',
        authorization: {
          typeUrl: '/cosmos.authz.v1beta1.GenericAuthorization',
          value: new Uint8Array([10, 30, 47, 99, 111, 115, 109, 111, 115]),
        },
        expiration: {
          seconds: BigInt(1735689600),
          nanos: 0,
        },
      },
      {
        granter: 'xion1granter...',
        grantee: 'xion1grantee2...',
        authorization: {
          typeUrl: '/cosmos.bank.v1beta1.SendAuthorization',
          value: new Uint8Array([10, 20, 10, 5, 117, 120, 105, 111, 110]),
        },
        expiration: {
          seconds: BigInt(1767225600),
          nanos: 0,
        },
      },
    ],
    pagination: {
      nextKey: new Uint8Array(),
      total: BigInt(2),
    },
  });

  console.log('4. Constructed Response:');
  console.log('   grants count:', response.grants.length);
  response.grants.forEach((grant, i) => {
    console.log(`\n   [${i}] GrantAuthorization:`);
    console.log(`       granter: ${grant.granter}`);
    console.log(`       grantee: ${grant.grantee}`);
    console.log(`       authorization.typeUrl: ${grant.authorization?.typeUrl}`);
  });
  console.log('\n   pagination.total:', response.pagination?.total?.toString());
  console.log('   typeUrl:', QueryGranterGrantsResponse.typeUrl);

  // 5. Encode/decode response
  const encodedResponse = QueryGranterGrantsResponse.encode(response).finish();
  const decodedResponse = QueryGranterGrantsResponse.decode(encodedResponse);
  console.log('\n5. Decoded grants count:', decodedResponse.grants.length);

  return { request, response };
}

// =============================================================================
// QueryGranteeGrantsRequest / Response
// =============================================================================
// List all grants received by a specific grantee.
// Endpoint: GET /xion/indexer/authz/v1/grants/grantee/{grantee}
// Type URL: /xion.indexer.authz.v1.QueryGranteeGrantsRequest

function queryGranteeGrantsExamples() {
  console.log('\n=== QueryGranteeGrants Examples ===\n');

  // 1. Construct request
  const request = QueryGranteeGrantsRequest.fromPartial({
    grantee: 'xion1grantee...', // List all grants TO this address
    pagination: {
      key: new Uint8Array(),
      offset: BigInt(0),
      limit: BigInt(20),
      countTotal: true,
    },
  });

  console.log('1. Constructed QueryGranteeGrantsRequest:');
  console.log('   grantee:', request.grantee);
  console.log('   pagination.limit:', request.pagination?.limit?.toString());
  console.log('   typeUrl:', QueryGranteeGrantsRequest.typeUrl);

  // 2. Encode request
  const encodedRequest = QueryGranteeGrantsRequest.encode(request).finish();
  console.log('\n2. Encoded request:', encodedRequest.length, 'bytes');

  // 3. ProtoMsg
  const protoMsg = QueryGranteeGrantsRequest.toProtoMsg(request);
  console.log('\n3. ProtoMsg typeUrl:', protoMsg.typeUrl);

  // === Response ===
  console.log('\n--- Response ---');

  const response = QueryGranteeGrantsResponse.fromPartial({
    grants: [
      {
        granter: 'xion1granter1...',
        grantee: 'xion1grantee...',
        authorization: {
          typeUrl: '/cosmos.authz.v1beta1.GenericAuthorization',
          value: new Uint8Array([10, 30, 47, 99, 111, 115, 109, 111, 115]),
        },
        expiration: {
          seconds: BigInt(1735689600),
          nanos: 0,
        },
      },
    ],
    pagination: {
      nextKey: new Uint8Array(),
      total: BigInt(1),
    },
  });

  console.log('4. Constructed Response:');
  console.log('   grants count:', response.grants.length);
  response.grants.forEach((grant, i) => {
    console.log(`\n   [${i}] GrantAuthorization:`);
    console.log(`       granter: ${grant.granter}`);
    console.log(`       grantee: ${grant.grantee}`);
    console.log(`       authorization.typeUrl: ${grant.authorization?.typeUrl}`);
  });
  console.log('\n   pagination.total:', response.pagination?.total?.toString());
  console.log('   typeUrl:', QueryGranteeGrantsResponse.typeUrl);

  // 5. Encode/decode response
  const encodedResponse = QueryGranteeGrantsResponse.encode(response).finish();
  const decodedResponse = QueryGranteeGrantsResponse.decode(encodedResponse);
  console.log('\n5. Decoded grants count:', decodedResponse.grants.length);

  return { request, response };
}

// =============================================================================
// Understanding Authz Grants
// =============================================================================

function authzExplanation() {
  console.log('\n=== Understanding Authz Grants ===\n');

  console.log('Authz (Authorization) allows one account to act on behalf of another.\n');

  console.log('Key Concepts:\n');
  console.log('  1. Granter: The account giving permission');
  console.log('  2. Grantee: The account receiving permission');
  console.log('  3. Authorization: What actions are permitted\n');

  console.log('Common Authorization Types:\n');
  const authTypes = [
    { type: 'GenericAuthorization', desc: 'Allows any message of a specific type' },
    { type: 'SendAuthorization', desc: 'Allows sending up to a spend limit' },
    { type: 'StakeAuthorization', desc: 'Allows staking operations (delegate, undelegate, redelegate)' },
  ];

  authTypes.forEach(({ type, desc }) => {
    console.log(`  • ${type}:`);
    console.log(`    ${desc}\n`);
  });

  console.log('Use Cases:\n');
  console.log('  • Fee grants: Let another account pay your fees');
  console.log('  • Automated staking: Let a bot manage your delegations');
  console.log('  • Multi-sig alternatives: Grant specific permissions to team members');
  console.log('  • Abstract accounts: Meta-transactions and gasless UX\n');

  console.log('Indexer vs Standard Authz:\n');
  console.log('  • This module provides OPTIMIZED QUERIES for grants');
  console.log('  • To CREATE grants, use: cosmos.authz.v1beta1.MsgGrant');
  console.log('  • To REVOKE grants, use: cosmos.authz.v1beta1.MsgRevoke');
  console.log('  • To EXECUTE with grant, use: cosmos.authz.v1beta1.MsgExec');
}

// =============================================================================
// Run All Examples
// =============================================================================

function main() {
  console.log('╔════════════════════════════════════════════════════════════╗');
  console.log('║  xion.indexer.authz.v1 Query Messages — Protobuf Examples ║');
  console.log('╚════════════════════════════════════════════════════════════╝\n');

  queryGrantsExamples();
  queryGranterGrantsExamples();
  queryGranteeGrantsExamples();
  authzExplanation();

  console.log('\n═══════════════════════════════════════════════════════════════');
  console.log('Type URL Reference (Requests):');
  console.log('  QueryGrantsRequest:        ', QueryGrantsRequest.typeUrl);
  console.log('  QueryGranterGrantsRequest: ', QueryGranterGrantsRequest.typeUrl);
  console.log('  QueryGranteeGrantsRequest: ', QueryGranteeGrantsRequest.typeUrl);
  console.log('\nType URL Reference (Responses):');
  console.log('  QueryGrantsResponse:        ', QueryGrantsResponse.typeUrl);
  console.log('  QueryGranterGrantsResponse: ', QueryGranterGrantsResponse.typeUrl);
  console.log('  QueryGranteeGrantsResponse: ', QueryGranteeGrantsResponse.typeUrl);
  console.log('\nREST Endpoints:');
  console.log('  GET /xion/indexer/authz/v1/grants');
  console.log('  GET /xion/indexer/authz/v1/grants/granter/{granter}');
  console.log('  GET /xion/indexer/authz/v1/grants/grantee/{grantee}');
  console.log('\nNote: This is an indexer module (query-only).');
  console.log('Use cosmos.authz.v1beta1 for grant/revoke transactions.');
  console.log('═══════════════════════════════════════════════════════════════');
}

// Execute if run directly
main();

export {
  queryGrantsExamples,
  queryGranterGrantsExamples,
  queryGranteeGrantsExamples,
  authzExplanation,
};

File: examples/kotlin/indexer-authz-v1-query-messages.kt
/**
 * xion.indexer.authz.v1 Query Messages — Protobuf Examples
 *
 * This file demonstrates how to construct, encode, and decode
 * query messages from the Authz Indexer Module (xion.indexer.authz.v1).
 *
 * Queries covered:
 * - QueryGrants: Get grants between specific granter and grantee
 * - QueryGranterGrants: List all grants issued by a granter
 * - QueryGranteeGrants: List all grants received by a grantee
 *
 * The Authz Indexer provides optimized queries for authorization grants,
 * allowing efficient lookup of who can act on behalf of whom.
 *
 * Note: This is an indexer module with query-only messages.
 * To create/revoke grants, use the standard cosmos.authz.v1beta1 module.
 */

package examples.kotlin

import xion.indexer.authz.v1.QueryOuterClass
import xion.indexer.authz.v1.queryGrantsRequest
import xion.indexer.authz.v1.queryGranterGrantsRequest
import xion.indexer.authz.v1.queryGranteeGrantsRequest
import cosmos.base.query.v1beta1.pageRequest
import com.google.protobuf.Any
import com.google.protobuf.Timestamp

// =============================================================================
// QueryGrantsRequest / Response
// =============================================================================
// Get grants between a specific granter and grantee.
// Endpoint: GET /xion/indexer/authz/v1/grants
// Type URL: /xion.indexer.authz.v1.QueryGrantsRequest

fun queryGrantsExamples() {
    println("=== QueryGrants Examples ===\n")

    val request = queryGrantsRequest {
        granter = "xion1granter..."
        grantee = "xion1grantee..."
        msgTypeUrl = "/cosmos.bank.v1beta1.MsgSend" // Optional: filter by message type
        pagination = pageRequest {
            limit = 10
            countTotal = true
        }
    }

    println("1. Constructed QueryGrantsRequest:")
    println("   granter: ${request.granter}")
    println("   grantee: ${request.grantee}")
    println("   msgTypeUrl: ${request.msgTypeUrl}")
    println("   pagination.limit: ${request.pagination.limit}")
    println("   typeUrl: /xion.indexer.authz.v1.QueryGrantsRequest")

    val encodedRequest = request.toByteArray()
    println("\n2. Encoded request: ${encodedRequest.size} bytes")

    // === Response ===
    println("\n--- Response ---")

    val response = QueryOuterClass.QueryGrantsResponse.newBuilder()
        .addGrants(QueryOuterClass.Grant.newBuilder()
            .setAuthorization(Any.newBuilder()
                .setTypeUrl("/cosmos.authz.v1beta1.GenericAuthorization")
                .setValue(com.google.protobuf.ByteString.copyFrom(byteArrayOf(10, 30, 47, 99, 111, 115, 109, 111, 115)))
                .build())
            .setExpiration(Timestamp.newBuilder()
                .setSeconds(1735689600L)
                .build())
            .build())
        .setPagination(cosmos.base.query.v1beta1.Pagination.PageResponse.newBuilder()
            .setTotal(1)
            .build())
        .build()

    println("3. Constructed Response:")
    println("   grants count: ${response.grantsCount}")
    response.grantsList.forEachIndexed { i, grant ->
        println("\n   [$i] Grant:")
        println("       authorization.typeUrl: ${grant.authorization.typeUrl}")
        println("       expiration: ${grant.expiration.seconds} (unix)")
    }
    println("\n   pagination.total: ${response.pagination.total}")
    println("   typeUrl: /xion.indexer.authz.v1.QueryGrantsResponse")
}

// =============================================================================
// QueryGranterGrantsRequest / Response
// =============================================================================
// List all grants issued by a specific granter.
// Endpoint: GET /xion/indexer/authz/v1/grants/granter/{granter}
// Type URL: /xion.indexer.authz.v1.QueryGranterGrantsRequest

fun queryGranterGrantsExamples() {
    println("\n=== QueryGranterGrants Examples ===\n")

    val request = queryGranterGrantsRequest {
        granter = "xion1granter..."
        pagination = pageRequest {
            limit = 20
            countTotal = true
        }
    }

    println("1. Constructed QueryGranterGrantsRequest:")
    println("   granter: ${request.granter}")
    println("   pagination.limit: ${request.pagination.limit}")
    println("   typeUrl: /xion.indexer.authz.v1.QueryGranterGrantsRequest")

    val encodedRequest = request.toByteArray()
    println("\n2. Encoded request: ${encodedRequest.size} bytes")

    // === Response ===
    println("\n--- Response ---")

    val response = QueryOuterClass.QueryGranterGrantsResponse.newBuilder()
        .addGrants(QueryOuterClass.GrantAuthorization.newBuilder()
            .setGranter("xion1granter...")
            .setGrantee("xion1grantee1...")
            .setAuthorization(Any.newBuilder()
                .setTypeUrl("/cosmos.authz.v1beta1.GenericAuthorization")
                .setValue(com.google.protobuf.ByteString.copyFrom(byteArrayOf(10, 30, 47, 99, 111, 115, 109, 111, 115)))
                .build())
            .setExpiration(Timestamp.newBuilder()
                .setSeconds(1735689600L)
                .build())
            .build())
        .addGrants(QueryOuterClass.GrantAuthorization.newBuilder()
            .setGranter("xion1granter...")
            .setGrantee("xion1grantee2...")
            .setAuthorization(Any.newBuilder()
                .setTypeUrl("/cosmos.bank.v1beta1.SendAuthorization")
                .setValue(com.google.protobuf.ByteString.copyFrom(byteArrayOf(10, 20, 10, 5, 117, 120, 105, 111, 110)))
                .build())
            .setExpiration(Timestamp.newBuilder()
                .setSeconds(1767225600L)
                .build())
            .build())
        .setPagination(cosmos.base.query.v1beta1.Pagination.PageResponse.newBuilder()
            .setTotal(2)
            .build())
        .build()

    println("3. Constructed Response:")
    println("   grants count: ${response.grantsCount}")
    response.grantsList.forEachIndexed { i, grant ->
        println("\n   [$i] GrantAuthorization:")
        println("       granter: ${grant.granter}")
        println("       grantee: ${grant.grantee}")
        println("       authorization.typeUrl: ${grant.authorization.typeUrl}")
    }
    println("\n   pagination.total: ${response.pagination.total}")
    println("   typeUrl: /xion.indexer.authz.v1.QueryGranterGrantsResponse")
}

// =============================================================================
// QueryGranteeGrantsRequest / Response
// =============================================================================
// List all grants received by a specific grantee.
// Endpoint: GET /xion/indexer/authz/v1/grants/grantee/{grantee}
// Type URL: /xion.indexer.authz.v1.QueryGranteeGrantsRequest

fun queryGranteeGrantsExamples() {
    println("\n=== QueryGranteeGrants Examples ===\n")

    val request = queryGranteeGrantsRequest {
        grantee = "xion1grantee..."
        pagination = pageRequest {
            limit = 20
            countTotal = true
        }
    }

    println("1. Constructed QueryGranteeGrantsRequest:")
    println("   grantee: ${request.grantee}")
    println("   pagination.limit: ${request.pagination.limit}")
    println("   typeUrl: /xion.indexer.authz.v1.QueryGranteeGrantsRequest")

    val encodedRequest = request.toByteArray()
    println("\n2. Encoded request: ${encodedRequest.size} bytes")

    // === Response ===
    println("\n--- Response ---")

    val response = QueryOuterClass.QueryGranteeGrantsResponse.newBuilder()
        .addGrants(QueryOuterClass.GrantAuthorization.newBuilder()
            .setGranter("xion1granter1...")
            .setGrantee("xion1grantee...")
            .setAuthorization(Any.newBuilder()
                .setTypeUrl("/cosmos.authz.v1beta1.GenericAuthorization")
                .setValue(com.google.protobuf.ByteString.copyFrom(byteArrayOf(10, 30, 47, 99, 111, 115, 109, 111, 115)))
                .build())
            .setExpiration(Timestamp.newBuilder()
                .setSeconds(1735689600L)
                .build())
            .build())
        .setPagination(cosmos.base.query.v1beta1.Pagination.PageResponse.newBuilder()
            .setTotal(1)
            .build())
        .build()

    println("3. Constructed Response:")
    println("   grants count: ${response.grantsCount}")
    response.grantsList.forEachIndexed { i, grant ->
        println("\n   [$i] GrantAuthorization:")
        println("       granter: ${grant.granter}")
        println("       grantee: ${grant.grantee}")
        println("       authorization.typeUrl: ${grant.authorization.typeUrl}")
    }
    println("\n   pagination.total: ${response.pagination.total}")
    println("   typeUrl: /xion.indexer.authz.v1.QueryGranteeGrantsResponse")
}

// =============================================================================
// Understanding Authz Grants
// =============================================================================

fun authzExplanation() {
    println("\n=== Understanding Authz Grants ===\n")

    println("Authz (Authorization) allows one account to act on behalf of another.\n")

    println("Key Concepts:\n")
    println("  1. Granter: The account giving permission")
    println("  2. Grantee: The account receiving permission")
    println("  3. Authorization: What actions are permitted\n")

    println("Common Authorization Types:\n")
    println("  • GenericAuthorization:")
    println("    Allows any message of a specific type\n")
    println("  • SendAuthorization:")
    println("    Allows sending up to a spend limit\n")
    println("  • StakeAuthorization:")
    println("    Allows staking operations (delegate, undelegate, redelegate)\n")

    println("Use Cases:\n")
    println("  • Fee grants: Let another account pay your fees")
    println("  • Automated staking: Let a bot manage your delegations")
    println("  • Multi-sig alternatives: Grant specific permissions to team members")
    println("  • Abstract accounts: Meta-transactions and gasless UX\n")

    println("Indexer vs Standard Authz:\n")
    println("  • This module provides OPTIMIZED QUERIES for grants")
    println("  • To CREATE grants, use: cosmos.authz.v1beta1.MsgGrant")
    println("  • To REVOKE grants, use: cosmos.authz.v1beta1.MsgRevoke")
    println("  • To EXECUTE with grant, use: cosmos.authz.v1beta1.MsgExec")
}

// =============================================================================
// Run All Examples
// =============================================================================

fun main() {
    println("╔════════════════════════════════════════════════════════════╗")
    println("║  xion.indexer.authz.v1 Query Messages — Protobuf Examples ║")
    println("╚════════════════════════════════════════════════════════════╝\n")

    queryGrantsExamples()
    queryGranterGrantsExamples()
    queryGranteeGrantsExamples()
    authzExplanation()

    println("\n═══════════════════════════════════════════════════════════════")
    println("Type URL Reference (Requests):")
    println("  QueryGrantsRequest:        /xion.indexer.authz.v1.QueryGrantsRequest")
    println("  QueryGranterGrantsRequest: /xion.indexer.authz.v1.QueryGranterGrantsRequest")
    println("  QueryGranteeGrantsRequest: /xion.indexer.authz.v1.QueryGranteeGrantsRequest")
    println("\nREST Endpoints:")
    println("  GET /xion/indexer/authz/v1/grants")
    println("  GET /xion/indexer/authz/v1/grants/granter/{granter}")
    println("  GET /xion/indexer/authz/v1/grants/grantee/{grantee}")
    println("\nNote: This is an indexer module (query-only).")
    println("Use cosmos.authz.v1beta1 for grant/revoke transactions.")
    println("═══════════════════════════════════════════════════════════════")
}
